statsmodels/miscmodels/robust_genericmle.py

```diff
@@ -0,0 +1,418 @@
+# -*- coding: utf-8 -*-
+"""
+
+Created on Thu Feb 27 13:32:36 2014
+
+Author: Josef Perktold
+"""
+
+import numpy as np
+from statsmodels.base.model import GenericLikelihoodModel
+from statsmodels.sandbox.regression import gmm
+
+import statsmodels.robust.norms as rnorms
+import statsmodels.robust.scale as rscale
+import statsmodels.robust.robust_linear_model as rlm
+from nltk.corpus.reader.rte import norm
+
+
+class MEstimator(GenericLikelihoodModel):
+
+    def __init__(self, *args, **kwds):
+        self.norm = rnorms.HuberT()
+        super(MEstimator, self).__init__(*args, **kwds)
+        self.scale_fixed = None
+
+    def transform_params(self, params):
+
+        if self.scale_fixed is None:
+            params_loc, scale = params[:-1], params[-1]
+            scale = np.abs(scale)
+        else:
+            params_loc = params
+            scale = self.scale_fixed
+
+        return params_loc, scale
+
+
+    def nloglikeobs(self, params):
+        # Note: norms is a loss function, so this is nloglikeobs
+
+        params_loc, scale = self.transform_params(params)
+
+        #nobs = len(self.endog)
+        resid_scaled = (self.endog - np.dot(self.exog, params_loc)) / scale
+        scale_fac = 0.491
+        rho_fac = scale * scale_fac
+        objective = (rho_fac * self.norm(resid_scaled / scale_fac) +
+                      rho_fac * np.log(scale * scale_fac))
+        return objective
+
+    def score(self, params):
+        return self.jac(params).sum(0)
+
+    def jac(self, params): #, **kwds):
+        # Todo   verify signs,  derivative of loglike or nloglike ?
+        params_loc, scale = self.transform_params(params)
+
+        resid_scaled = (self.endog - np.dot(self.exog, params_loc)) / scale / 0.491
+        score_obs = -self.norm.psi(resid_scaled)[:,None] * self.exog #/ scale
+        # note: extra scale cancels between nominator and denominator
+
+        if self.scale_fixed is None:
+            def func(s):
+                s = np.atleast_1d(s)
+                #for derivative wrt. scale
+                p = np.concatenate((params_loc, s))
+                return self.loglikeobs(p)
+
+            import statsmodels.tools.numdiff as nd
+            grad_scale = nd.approx_fprime(np.array([scale]), func)
+            return -np.column_stack((score_obs, -grad_scale))
+
+        else:
+            return -score_obs
+
+    def predict(self, params, exog=None):
+        if exog is None:
+            exog = self.exog
+        return np.dot(exog, params[:self.exog.shape[1]])
+
+
+class MEstimatorHD(GenericLikelihoodModel):
+
+    def __init__(self, *args, **kwds):
+        self.norm = kwds.get('norm', rnorms.HuberT())
+        super(MEstimatorHD, self).__init__(*args, **kwds)
+        self.scale_fixed = None
+        # TODO verify scale_fac
+        # stats.norm.expect(lambda t: t*norm.psi(t) - norm.rho(t))
+        self.scale_fac = 0.35508
```

Comment:
another scale_fac with different value ?

Note moment condition for s for HD differs from log(s) mle-like version
so fisher consistency requires different expectation